docs/developer/releasing.md

```diff
@@ -113,25 +124,55 @@ For patch releases, the release shepard will:
 
 3. Create the patch release tag.
 
-After the release shepard pushes a new tag, they must [publish the release](#publishing-the-release).
+4. [Publish the release](#publishing-the-release)
 
 ### Updating code
 
-The codebase must be updated whenever a new release is being made to reference
-the upcoming release tag:
+The codebase must be updated whenever a new release is being made to reference the upcoming release tag.
 
-* Modify `CHANGELOG.md` with a new version number and its release date.
-* Go through the entire repository and find references to the previous release
-  version, updating them to reference the new version.
-* Run `make generate-manifests` and `make generate-dashboards` to update
-  manifests in case they are stale.
+#### Update the changelog
 
 NOTE: Any time CHANGELOG.md is updated for a release, it should first be done
 via PR to the release branch, and then by a second PR to main.
 
+When creating a release **candidate** (an `-rc.N` version):
+
+* Add a new section in the changelog for the new release candidate and include 
+today's date. E.g. "v0.31.0-rc.0 (2023-01-26)".
+* All items form the "unreleased" section will move to a new section for the upcoming release version.
+* See [here](https://github.com/grafana/agent/pull/2838/files) for an example pull request.
+* Sanity check that the changelog doesn't contain features which don't belong to it. Sometimes features get added to the wrong version in the changelog due to bad merges.
+
+When creating a **stable** release:
+
+* Replace the version from the release candidate section to be the stable release instead 
+of the `-rc.N` version.
+* We do not leave release candidates in the changelog once there is a stable version for them.
+* Make sure to also update the **date** in the title.
+* See [here](https://github.com/grafana/agent/pull/2873/files) for an example pull request.
+
+#### Search and replace the old version with the new one
+
+Go through the entire repository and find references to the previous release
+  version, updating them to reference the new version **where necessary**.
+
+NOTE: There are files such as 
+"[pkg/operator/defaults.go](https://github.com/grafana/agent/blob/main/pkg/operator/defaults.go)", 
+where the version sometimes should not be replaced but added to a list of versions.
+At the time of this writing, `defaults.go` is the only such file.
+
+#### Update defaults.go
+
+Add the new version to the "[pkg/operator/defaults.go](https://github.com/grafana/agent/blob/main/pkg/operator/defaults.go)" file. If there is a release candidate (`-rc.N`) version, remove it.
+
+#### Update manifests and dashboards
+
+Run `make generate-manifests` and `make generate-dashboards` to update
+manifests in case they are stale.
```

Comment:
Ideally we remove this from the release instructions and instead run `make generate-manifests generate-dashboards` in CI and fail a PR which causes them to be stale.

Reply:
If I understand you correctly, what these two `make` commands do is the same as what we would do by search and replace of the version through the codebase? Hence we can do search and replace for now as our release process, and use the `make` commands just to double check that we replaced everything correctly?

I am not too sure how to make sure those `make` commands completed successfully. Maybe the best way would be to have the makefile return an error code which causes Drone to fail the CI? Here I am guessing that Drone would fail when one of the `commands` in the docker pipeline returns a non-zero code.

But then, how would the makefile know that the command failed? I ran a `tk show` command just to see what happens when `tk` exits with an error, and the return code was 0. A 0 normally indicates success. Maybe there is a way other than error codes to check for failures?

If you want to, I could move this into a separate issue and could mention that we should remove the steps from the documentation when this is added to the CI?

Reply:
I'm suggesting that we have a CI job to make sure that the manifests are never stale. We do this for Helm chart tests: https://github.com/grafana/agent/blob/9c068ac3993983b1e3160af6e9af117fdda35f64/.github/workflows/helm-test.yml#L34-L41

That way, there would be no reason to include this step in the release step instructions because manifests would not be stale at release time.

Reply:
Ok, thank you, I'll delete the instructions from here and will make sure to update the CI on the other [PR](https://github.com/grafana/agent/pull/2966).